### PR TITLE
nixos/unit: enable sanboxing

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Enable sanboxing.

cc @aanderse @flokli 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: nixos/modules/services/web-servers/unit/default.nix [hunk 1]
```diff
@@ -99,17 +99,34 @@ in {
...
+        # Capabilities
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" "CAP_SETGID" "CAP_SETUID" ];
+        # Security
+        NoNewPrivileges = true;
+        # Sanboxing
+        ProtectSystem = "full";
+        ProtectHome = true;
         RuntimeDirectory = "unit";
         RuntimeDirectoryMode = "0750";
+        RuntimeDirectoryPreserve = "yes";
```
Comment (flokli) [inline nixos/modules/services/web-servers/unit/default.nix]: Why do we set `RuntimeDirectoryPreserve=yes ? Looking at the `preStart` script, the only thing we do in `/run` is remove a file unit might have created during runtime…

Comment (Izorkin) [reply nixos/modules/services/web-servers/unit/default.nix]: During debugging, sometimes an error occurs:
```
дек 08 23:03:23 NixOS-Test unitd[8140]: 2019/12/08 23:03:23 [info] 8140#8140 unit started
дек 08 23:03:23 NixOS-Test unitd[8140]: 2019/12/08 23:03:23 [alert] 8140#8140 unlink("/run/unit/unit.pid") failed (2: No such file or directory)
дек 08 23:03:23 NixOS-Test unitd[8140]: ^[[0;1;39m[alert] 8140#8140 unlink("/run/unit/unit.pid") failed (2: No such file or directory)
дек 08 23:03:23 NixOS-Test systemd[1]: ^[[0;1;39m^[[0;1;39munit.service: Main process exited, code=exited, status=1/FAILURE
```
As a result, it is not clear - the process can`t create a folder `/run/unit`, or the problem is different

Comment (flokli) [reply nixos/modules/services/web-servers/unit/default.nix]: `/run/unit` should be created via the `RuntimeDirectory=unit` directive.

The error is about `unit` trying to remove `/run/unit/unit.pid`. We should disable pid file handling entirely if possible.

File: nixos/modules/services/web-servers/unit/default.nix [hunk 2]
```diff
@@ -94,6 +94,7 @@ in {
       path = with pkgs; [ curl ];
       preStart = ''
         test -f '/run/unit/control.unit.sock' || rm -f '/run/unit/control.unit.sock'
+        test -f '${cfg.stateDir}/conf.json'   || rm -f '${cfg.stateDir}/conf.json'
```
Comment (flokli) [inline nixos/modules/services/web-servers/unit/default.nix]: This looks dangerous. What's written into `conf.json`? Why does it end in `${cfg.stateDir}` in first place, if we don't want it?

Comment (Izorkin) [reply nixos/modules/services/web-servers/unit/default.nix]: In `conf.json` last working configuration is stored.
At the `postStart` stage, the current configuration is loaded.

File: nixos/modules/services/web-servers/unit/default.nix (hunk 1 above)
Comment (Izorkin) [reply nixos/modules/services/web-servers/unit/default.nix]: Removed `RuntimeDirectoryPreserve`.
The unit process automatically creates a pid file, not systemd creates.

File: nixos/modules/services/web-servers/unit/default.nix [hunk 3]
```diff
@@ -94,22 +94,39 @@ in {
       path = with pkgs; [ curl ];
       preStart = ''
         test -f '/run/unit/control.unit.sock' || rm -f '/run/unit/control.unit.sock'
```
Comment (flokli) [inline nixos/modules/services/web-servers/unit/default.nix]: We can stop removing stuff in `/run/unit/*`, now that `RuntimeDirectoryPreserve` isn't set anymore, the directory is empty during start.

Comment (Izorkin) [reply nixos/modules/services/web-servers/unit/default.nix]: Removed

**Pull request merged**